### PR TITLE
Fused Cross Entropy Triton - Loss Scaling and Vanishing Grads Bugfix

### PR DESCRIPTION
**Pull request by sarthak-amd**
# Description

The Fused Cross Entropy Triton Kernel currently has 2 bugs
1. if `ignore_idx` is not None`, the loss should be computed only over valid tokens and not all tokens (new fix)
2. gradient scaling when `reduce_loss=False`. (This is already fixed in upstream)
    - If `reduced loss=False`, we should compute per token loss and not reduce it else it would shrink the gradients by 1/N giving wrong (higher) loss. 
    - if `reduce_loss=False`, `grad_output` is a tensor, not a scalar. We need to load 1 value per row instead of just a scalar. 

This fix is validated on Llama3.1 8B model for Pre-training. 

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by wenchenvincent** [comment]
@sarthak-amd  Could you post the PR for the upstream fix?

File: transformer_engine/pytorch/cross_entropy.py [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# This file was modified for portability to AMDGPU
```
Comment (wenchenvincent) [inline transformer_engine/pytorch/cross_entropy.py]: There is no real change in this file. Let's keep this file intact and then we don't need to add the AMD copyright statement.

**Comment by sarthak-amd** [comment]
> @sarthak-amd Could you post the PR for the upstream fix?

https://github.com/NVIDIA/TransformerEngine/commit/e9a5fa4e368464f3b310b90ab7f670f35319344b @wenchenvincent

**Comment by wenchenvincent** [comment]
> > @sarthak-amd Could you post the PR for the upstream fix?
> 
> [NVIDIA/TransformerEngine@e9a5fa4](https://github.com/NVIDIA/TransformerEngine/commit/e9a5fa4e368464f3b310b90ab7f670f35319344b) @wenchenvincent

Another fix came from the upstream PR https://github.com/NVIDIA/TransformerEngine/pull/1879. Is the change of test in that PR also reflected?

**Comment by wenchenvincent** [comment]
For the fix for `ignore_idx`, is there a test for it (without the fix, the test would fail)?

**Review by wenchenvincent — CHANGES_REQUESTED** [review]
@sarthak-amd Could you refactor the PR as 3 commits:
- 2 commits would be cherrypicking from the upstream PRs.
- 1 commit for the `ignore_idx` with a test to cover it.

This way the PR would be very clear and easy to understand.

**Comment by wenchenvincent** [comment]
@sarthak-amd Could you address the comments? Also, please rebase upon latest dev so that hot fixes for sgpu tests could pass.

**Comment by wenchenvincent** [comment]
@sarthak-amd Could you rebase the PR and update it per reviewer comments?
